### PR TITLE
Add more metrics to categorical.py, modified the wrappers.py the support Weibull’s plotting position and add new file implement_categorical_metrics.ipynb to provide tutorial

### PR DESCRIPTION
**Pull request by LQscience**
# Modified:   weatherbenchX/metrics/categorical.py
The following metrics are added: 
- Symmetric extreme dependence score (SEDS)
- Symmetric extremal dependence index (SEDI)
- False alarm rate (F), probability of false detection
- Heidke skill score (HSS), Cohen’s Kappa 
- Odds ratio skill score (ORSS), Yule’s Q
- Extreme dependence score (EDS)
- Extremal dependence index (EDI)
- Gilbert skill score (GSS), equitable threat score
- Peirce’s skill score (PSS), Hanssen and Kuipers discriminant
- Specificity, true negative rate (TNR)
- Area under receiver operating characteristic (ROC) curve (AUC)
- ROC skill score (ROCSS)
- Relative economic value (REV)

# Modified:   weatherbenchX/metrics/wrappers.py
- modified the function "binarize_thresholds" to support threshold of Xarray.DataArray
- Add new class "EnsembleToProbabilistic" to convert ensemble forecasts into probabilitic forecast using the Weibull’s plotting position (Makkonen, 2006).
Makkonen, L.: Plotting Positions in Extreme Value Analysis, Journal of Applied Meteorology and Climatology, 
      45, 334–340, https://doi.org/10.1175/JAM2349.1, 2006.

# Add new file: docs/source/how_to/implement_categorical_metrics.ipynb
This notebook provides tutorial to use these categorical metrics.
Also, it is a part of code to reproduce the analysis of the manuscript "The extension of the WeatherBench 2 to binary hydroclimatic forecasts" (https://egusphere.copernicus.org/preprints/2025/egusphere-2025-3/) that has been submitted to the Geoscientific Model Development.

### CODE REVIEW

**Comment by raspstephan** [comment]
Hi,
thank you so much for this PR. I will take a close look tomorrow.

**Comment by raspstephan** [comment]
This is really great work and will be a valuable addition to WeatherBench. Generally, the added code looks great. I have a few suggestions though:

1. Since the notebook is relatively specific for your paper, I would suggest adding the notebook as a supplement to the paper instead of adding it to the WB-X documentation. Instead, I can add a section to the documentation called something like "Used by" which lists all example of people using WB-X and add a reference to your paper and notebook there.

2. Adding an option of using a DataArray as a threshold is great and is actually something we were planning to do as well. However, currently this only allows using a DataArray which means that the same thresholds apply to all variables. Some users might want to use different thresholds for different variables though (I think you do as well judging from your notebook.) One way to do this would be to also allow using an xr.Dataset and adding a logic something like (I did not actually test this, just pseudo-code):
```
if isinstance(thresholds, xr.Dataset):
  threshold_for_variable = thresholds[x.name]
```

3. Since this is quite a lot of new functionality and metrics, I would love it if you could add some tests. Specifically for the new xr.DataArray/Dataset options in the wrapper. You can decide for yourself whether you'd like to add tests for the metrics.

File: weatherbenchX/metrics/categorical.py [hunk 1]
```diff
@@ -487,3 +487,717 @@ def _values_from_mean_statistics_with_internal_names(
...
+        statistic_values['TruePositives'] + statistic_values['FalseNegatives'])
+
+    return (
+      np.log(forecast_rate) - np.log(hit_rate)
+    ) / (
+       np.log(base_rate) + np.log(hit_rate)
+    )
+
+
+class F(base.PerVariableMetric):
```
Comment (raspstephan) [inline weatherbenchX/metrics/categorical.py:697]: I think it would be better if this was named FalseAlarmRate since F can be a little confusing.

File: weatherbenchX/metrics/categorical.py [hunk 2]
```diff
@@ -487,3 +487,717 @@ def _values_from_mean_statistics_with_internal_names(
...
+  def _compute_per_variable(
+      self,
+      predictions: xr.DataArray,
+      targets: xr.DataArray,
+  ) -> xr.DataArray:
+
+    return ((predictions - targets) ** 2).astype(np.float32)
+
+
+class BrierScore(base.PerVariableMetric):
```
Comment (raspstephan) [inline weatherbenchX/metrics/categorical.py:1168]: So Brier Score actually doesn't require any additional code since mathematically it's the same as deterministic.MSE(). However, I would be happy to still implement it (but then using the squared error definition from deterministic.py).

But could you explain why you need to add a reduce_dims argument? This should be handled by the aggregator, no?

File: weatherbenchX/metrics/categorical.py [hunk 3]
```diff
@@ -487,3 +487,717 @@ def _values_from_mean_statistics_with_internal_names(
...
+    e_clim = event_ratio.where(event_ratio < self._cost_loss_ratios, self._cost_loss_ratios)
+    e_fcst = (
+        statistic_values['TruePositives'] + statistic_values['FalsePositives']
+      ) / event_nonevent_count * self._cost_loss_ratios + statistic_values['FalseNegatives'] / event_nonevent_count
+    e_perfect = event_ratio * self._cost_loss_ratios
+    
+    return ((e_clim - e_fcst) / (e_clim - e_perfect)).max(dim=self._threshold_probability_dim)
+
+
+class SquaredError(base.PerVariableStatistic):
```
Comment (raspstephan) [inline weatherbenchX/metrics/categorical.py:1152]: Same as deterministic.SquaredError, see comment below.

File: weatherbenchX/metrics/categorical.py [hunk 4]
```diff
@@ -487,3 +487,717 @@ def _values_from_mean_statistics_with_internal_names(
...
+        statistic_values['FalsePositives'] + statistic_values['TrueNegatives']
+    )
+    return (
+      np.log(false_alarm_rate) - np.log(hit_rate) + np.log(1-hit_rate) - np.log(1-false_alarm_rate)
+    ) / (
+       np.log(false_alarm_rate) + np.log(hit_rate) + np.log(1-hit_rate) + np.log(1-false_alarm_rate)
+    )
+
+
+class AUC(base.PerVariableMetric):
```
Comment (raspstephan) [inline weatherbenchX/metrics/categorical.py:899]: I think here a test would be great to see if the function returns the expected results in some edge cases.

File: weatherbenchX/metrics/categorical.py (hunk 2 above)
Comment (mjwillson) [reply weatherbenchX/metrics/categorical.py:1168]: +1 perhaps just aliasing deterministic.MSE here as BrierScore would be sufficient?

You add a float32 cast here vs in deterministic.SquaredError, but I'm not sure this is necessary if this is wrapped with wrappers.ContinuousToBinary as binarize_thresholds already returns floats, due to the .where: https://github.com/google-research/weatherbenchX/blob/main/weatherbenchX/metrics/wrappers.py#L69. (@raspstephan I think the docs on this are incorrect, AFAICT it always returns floats regardless whether NaNs are present or not?).

Also note that treating Brier score as a special case of MSE allows you to reuse probabilistic.UnbiasedEnsembleMeanSquaredError to get a fair / unbiased estimate of Brier score from an ensemble. (In the special case of Brier score this is equivalent to the unbiased estimator from Ferro, C.A., 2007: Comparing Probabilistic Forecasting Systems with the Brier Score, Eqn 8 with M->inf)

File: weatherbenchX/metrics/categorical.py [hunk 5]
```diff
@@ -487,3 +487,717 @@ def _values_from_mean_statistics_with_internal_names(
...
+    }
+
+  def _values_from_mean_statistics_per_variable(
+      self,
+      statistic_values: Mapping[Hashable, xr.DataArray],
+  ) -> xr.DataArray:
+    """Computes metrics from aggregated statistics."""
+
+    event_count = statistic_values['TruePositives'] + statistic_values['FalseNegatives']
+    event_nonevent_count = event_count + statistic_values['FalsePositives'] + statistic_values['TrueNegatives']
```
Comment (mjwillson) [inline weatherbenchX/metrics/categorical.py:1071]: FYI this should always be 1 (up to numerical error) because weatherbench gives you *means* of the statistics here. The entries of the confusion matrix always sum to 1 on a per-example basis, and so their means should sum to 1 as well.

In the case of numerical error I'm actually not 100% sure what's best to do here, but I don't think it would generally make a big difference so perhaps better to go with the simpler option and avoid the extra division?

File: weatherbenchX/metrics/categorical.py [hunk 6]
```diff
@@ -487,3 +487,717 @@ def _values_from_mean_statistics_with_internal_names(
...
+    event_ratio = event_count / event_nonevent_count
+    
+    e_clim = event_ratio.where(event_ratio < self._cost_loss_ratios, self._cost_loss_ratios)
+    e_fcst = (
+        statistic_values['TruePositives'] + statistic_values['FalsePositives']
+      ) / event_nonevent_count * self._cost_loss_ratios + statistic_values['FalseNegatives'] / event_nonevent_count
+    e_perfect = event_ratio * self._cost_loss_ratios
+    
+    return ((e_clim - e_fcst) / (e_clim - e_perfect)).max(dim=self._threshold_probability_dim)
```
Comment (mjwillson) [inline weatherbenchX/metrics/categorical.py:1149]: Most of this duplicates code in REV._values_from_mean_statistics_per_variable, would be nice to factor out into a shared helper.

File: weatherbenchX/metrics/categorical.py [hunk 7]
```diff
@@ -487,3 +487,717 @@ def _values_from_mean_statistics_with_internal_names(
...
+  e_perfect = (TP + FN) / total * cost_loss_ratios
+
+  REV = (e_clim - e_fcst) / (e_clim - e_perfect)
+
+  PREV = max(REV) for threshold_probability in [0, 1]
+  """
+  def __init__(
+      self,
+      cost_loss_ratios: Union[float, Sequence[float]],
+      threshold_probability_dim: str,
```
Comment (mjwillson) [inline weatherbenchX/metrics/categorical.py:1106]: Could you document a bit on the requirements for this threshold_probability_dim and how it would typically be supplied? I think the idea is that you wrap this with ContinuousToBinary with multiple thresholds set for the probability, right?

File: weatherbenchX/metrics/categorical.py [hunk 8]
```diff
@@ -487,3 +487,717 @@ def _values_from_mean_statistics_with_internal_names(
...
+  ) -> xr.DataArray:
+    """Computes metrics from aggregated statistics."""
+    assert self._threshold_probability_dim in statistic_values['TruePositives'].dims, (
+        f"Threshold probability dim {self._threshold_probability_dim} not found in "
+        f"TruePositives statistic"
+    )
+    hit_rate = statistic_values['TruePositives'] / (
+        statistic_values['TruePositives'] + statistic_values['FalseNegatives']
+    )
+    # print(hit_rate)
```
Comment (mjwillson) [inline weatherbenchX/metrics/categorical.py:1003]: Please remove, here and below

File: weatherbenchX/metrics/categorical.py [hunk 9]
```diff
@@ -487,3 +487,717 @@ def _values_from_mean_statistics_with_internal_names(
...
+    false_alarm_rate = statistic_values['FalsePositives'] / (
+        statistic_values['FalsePositives'] + statistic_values['TrueNegatives']
+    )
+    # print(false_alarm_rate)
+    auc_dims = tuple(set(hit_rate.dims) - {self._threshold_probability_dim})
```
Comment (mjwillson) [inline weatherbenchX/metrics/categorical.py:1008]: This shares most of its code with AUC, could we please factor out the shared code into a helper?

Comment (mjwillson) [reply weatherbenchX/metrics/categorical.py:1008]: Alternatively I think it would be fair to decide to implement just one of AUC and ROCSS = 2 * AUC - 1, since they can trivially be transformed into eachother by the end-user of the metrics. Maybe just AUC as I think that's better known?

File: weatherbenchX/metrics/categorical.py (hunk 1 above)
Comment (mjwillson) [reply weatherbenchX/metrics/categorical.py:697]: +1. We have a lot of dense acronyms in class names here. I think expanding the acronyms would generally be good for readability, except  in cases where this would lead to a very long name, or where the acronym is very well-known and unambiguous.

So e.g. SEEPS -> StableEquitableErrorInProbabilitySpaceScore would probably be overkill, but things like F -> FalsePositiveRate or PSS -> PierceSkillScore seem like an improvement to me.

**Review by mjwillson — COMMENTED** [review]
Thanks, this is a really great contribution! Made a few comments, mostly minor.

File: weatherbenchX/metrics/wrappers.py [hunk 10]
```diff
@@ -395,3 +401,31 @@ def _values_from_mean_statistics_with_internal_names(
     return self.metric._values_from_mean_statistics_with_internal_names(  # pylint: disable=protected-access
         statistic_values
     )
+
+
+class EnsembleToProbabilistic(InputTransform):
+  """
+  Convert ensemble forecasts into probabilitic forecast using the Weibull’s plotting position (Makkonen, 2006).
```
[the remark this answers is not in the document]
Comment (mjwillson) [reply weatherbenchX/metrics/wrappers.py:408]: Also good to clarify that you need to binarize forecasts before this (e.g. by first wrapping with ContinuousToBinary)

**Comment by LQscience** [comment]
> Thanks, this is a really great contribution! Made a few comments, mostly minor.

Thanks you very much  for the positive feedback and detailed suggestions! I'll check and address all the issues this weekend.

File: weatherbenchX/metrics/wrappers.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (LQscience) [reply weatherbenchX/metrics/wrappers.py:408]: Thank you for the constructive comments. The class name has been modified to "WeibullEnsembleToProbabilistic". In the meantime, the need to binarize the forecasts before this wrapper has been clrified in the introduction.

File: weatherbenchX/metrics/wrappers.py [hunk 11]
```diff
@@ -395,3 +401,31 @@ def _values_from_mean_statistics_with_internal_names(
...
+    super().__init__(which)
+    self._ensemble_dim = ensemble_dim
+    self._skipna = skipna
+  @property
+  def unique_name_suffix(self) -> str:
+    return 'ensemble_to_probabilistic'
+
+  def tranform_fn(self, da: xr.DataArray) -> xr.DataArray:
+    ensemble_members = da[self._ensemble_dim].size
+    return da.sum(self._ensemble_dim, skipna=self._skipna)/(ensemble_members+1)
```
[the remark this answers is not in the document]
Comment (LQscience) [reply weatherbenchX/metrics/wrappers.py:431]: I have realized it is true that the ensemble_members would generally need to be adjusted based on the number of non-NaN ensemble members. So the skipna is expected to be False here and I have modified that.

File: weatherbenchX/metrics/categorical.py (hunk 8 above)
Comment (LQscience) [reply weatherbenchX/metrics/categorical.py:1003]: Sorry for the lines not deleted after testing. I have removed them.

File: weatherbenchX/metrics/categorical.py (hunk 9 above)
Comment (LQscience) [reply weatherbenchX/metrics/categorical.py:1008]: Thanks for your suggestion. I have deleted the ROCSS.

File: weatherbenchX/metrics/categorical.py (hunk 1 above)
Comment (LQscience) [reply weatherbenchX/metrics/categorical.py:697]: Sorry for the confusion. I have checked all metrics and modified their names to their full names.

File: weatherbenchX/metrics/categorical.py (hunk 7 above)
Comment (LQscience) [reply weatherbenchX/metrics/categorical.py:1106]: I have to clarify that the "threshold_probability_dim" here is different with the "threshold_dim" of the "wrappers.ContinuousToBinary". The former is used to convert the probabilistic forecasts into binary forecasts, while the latter is used to convert the continuous forecasts into binary forecasts. For calculating the PREV, we need to predefined a set of probability thresholds over which the maximum REV is calculated.

File: weatherbenchX/metrics/categorical.py (hunk 6 above)
Comment (LQscience) [reply weatherbenchX/metrics/categorical.py:1149]: As the PREV can be easily calculated using the max function, i decided to delete the PREV class.

File: weatherbenchX/metrics/categorical.py (hunk 2 above)
Comment (LQscience) [reply weatherbenchX/metrics/categorical.py:1168]: The ".astype(np.float32)" is added here because I found that the calculation for global grid cells ususlly be out of memory for my PC. Of course,  it can be deleted.

File: weatherbenchX/metrics/categorical.py (hunk 4 above)
Comment (LQscience) [reply weatherbenchX/metrics/categorical.py:899]: Thank you for your suggestion. I decide to remove this function in this PR and add it in the next PR when I complete to prapare the "metrics_test.py".

File: weatherbenchX/metrics/categorical.py (hunk 3 above)
Comment (LQscience) [reply weatherbenchX/metrics/categorical.py:1152]: Thank you for your suggestion. I decide to remove this function in this PR and add it in the next PR when I complete to prapare the "metrics_test.py".

File: weatherbenchX/metrics/categorical.py (hunk 2 above)
Comment (LQscience) [reply weatherbenchX/metrics/categorical.py:1168]: Thank you for your suggestion. I decide to remove this function in this PR and add it in the next PR when I finish to prapare the "metrics_test.py".

File: weatherbenchX/metrics/wrappers.py [hunk 12]
```diff
@@ -63,15 +63,15 @@ def binarize_thresholds(
   Returns:
     binary_x: Binarized DataArray.
   """
-  if not isinstance(thresholds, xr.DataArray):
+  if not isinstance(thresholds, xr.Dataset):
```
Comment (langmore) [inline weatherbenchX/metrics/wrappers.py:66]: In `binarize_thresholds` def, you can update the typing of thresholds to `Iterable[float] | xr.Dataset`

File: ._.DS_Store
Comment (raspstephan) [inline ._.DS_Store:1]: Please remove :)

**Comment by raspstephan** [comment]
Thanks a lot for addressing all the comments. I think the only thing that is remaining is to add tests. I think we definitely want to test the behavior or the wrapper to make sure using an xr.Dataset as a threshold works as expected. It would be great to also have tests for the new metrics but we don't have full test coverage of the existing metrics, so it's not a strict requirement.

**Comment by LQscience** [comment]
> Thanks a lot for addressing all the comments. I think the only thing that is remaining is to add tests. I think we definitely want to test the behavior or the wrapper to make sure using an xr.Dataset as a threshold works as expected. It would be great to also have tests for the new metrics but we don't have full test coverage of the existing metrics, so it's not a strict requirement.

Thank you very much for your positive feedback. I have already test the added metrics and will test several other complex metrics in the following few days. After testing, I will submit these complex metrics as well as the test of all the metrics.

File: ._.DS_Store
Comment (LQscience) [reply ._.DS_Store:1]: OK, I have already removed it.

**Comment by langmore** [comment]
I have a pending (internal) change that adds a file for tests of `InputTransform`. This may be a good home for some of your tests. In particular, I'm encouraging users to test the `InputTransform` *without* any metric.  So, @LQscience if you're writing tests of `InputTransform`'s now, just add them to `metrics_test.py` (we can move later), and test the transform without a metric.

**Comment by LQscience** [comment]
> I have a pending (internal) change that adds a file for tests of `InputTransform`. This may be a good home for some of your tests. In particular, I'm encouraging users to test the `InputTransform` _without_ any metric. So, @LQscience if you're writing tests of `InputTransform`'s now, just add them to `metrics_test.py` (we can move later), and test the transform without a metric.

Thanks for the kind reminder! I’ll add the tests to "metrics_test.py" as soon as possible.

**Comment by raspstephan** [comment]
Hey @LQscience, just checking in whether you are still working on this? No rush but some of us are quite excited for the features you implemented, so if I can help with anything let me know.

**Comment by LQscience** [comment]
> Hey @LQscience, just checking in whether you are still working on this? No rush but some of us are quite excited for the features you implemented, so if I can help with anything let me know.

Dear Rasp,

I sincerely apologize for the delay in my response. I have been preoccupied with other tasks recently, which prevented me from addressing your feedback in a timely manner. I truly appreciate your patience and understanding.

I have now made the following updates based on your feedback:
1. Improve the `binarize_thresholds` and `ContinuousToBinary` functions in the `wrappers.py`to support threshold in the forms of  xarray.DataArray and xarray.Dataset. 
2. Add a new class `WeibullEnsembleToProbabilistic` into the `wrappers.py` for converting ensemble forecasts into probabilistic forecasts using Weibull’s plotting position. 
3. Add tests in the `wrappers_test.py` to validate these changes above.

For more concise and efficient modification and review processes, I will close this lengthy pull request. Please refer to the new [pull request](https://github.com/google-research/weatherbenchX/pull/52) for the changes above. I am currently continuing to improve and test the remaining issues and will make new pull request.

Please let me know if there are any further modifications needed. Thank you for your guidance and support.

Best regards,
Qiang Li

**Comment by raspstephan** [comment]
Just out of curiosity, are you planning to send another PR with REV?
